### ISSUE TITLE
Dispatching action using @Select sequence requires setTimeout

### ISSUE THREAD

**Issue by wheredoesyourmindgo** [ngxs/store#103]
How come I need to use setTimeout() to properly dispatch an action when using a Select. [See stackblitz example](https://stackblitz.com/edit/ngxs-simple-en4y8f).  In the example, if the setTimeout is omitted, the dispatch happens too quickly for lack of a better explanation and the view doesn't continually show Odd numbers.  Is this a bad practice? Should I be returning a dispatched event in the action instead, as illustrated in [this example](https://amcdnl.gitbooks.io/ngxs/concepts/state.html#dispatching-actions-from-actions)? Thanks in advance.

**Comment by deebloo** [comment]
@wheredoesyourmindgo what is it that you are trying to accomplish?

**Comment by wheredoesyourmindgo** [comment]
In this trivial example, the app component uses a Select decorator to grab a slice of the state, and assigns it to a property in the AppComponent class called `count$`. The template is displaying the current value of count$ using async pipe. When subscribing to the `count$` observable during `OnInit()` I check to see if the number is an even number, if it is, Dispatch is called on the Add Action and increases the value by one. I expect that the view will always display an odd number. If I enclose the the dispatch action call in a `setTimeout()` function the view always shows odd numbers (as expected), however, when the dispatch action call happens outside of the scope of the `setTimeout` function, the view always shows even numbers. This can be illustrated in this example by simply commenting lines 22 & 24 in `app.component.ts`. Thanks @deebloo and @amcdnl

**Comment by yarrgh** [comment]
Works if you change the logic in `ngOnInit()` to:
```typescript
  ngOnInit(): void {
    this.count$.pipe(distinctUntilChanged()).subscribe(v => {
      if (v % 2 !== 0) {
        // setTimeout(() => {
          this.store.dispatch(new Add());
        // });
      }
    });
  }
```

Although it starts off at 0 instead of 1.

**Comment by yarrgh** [comment]
It does what it should (mostly at least). If you stick a `console.log(v)` before the if statement you'll see that the correct value does show up. However it doesn't make it to the DOM for some reason. Value changing too quickly for change detection?

```typescript
  ngOnInit(): void {
    this.count$.pipe(distinctUntilChanged()).subscribe(v => {
        console.log(v);
      if (v % 2 === 0) {
        // setTimeout(() => {
          this.store.dispatch(new Add());
        // });
      }
    });
  }
```
![image](https://user-images.githubusercontent.com/1577822/37680196-c88f835c-2c48-11e8-82ae-360ce3682996.png)

**Comment by wheredoesyourmindgo** [comment]
@yarrgh Correct. The correct value may be saved in State and shown correctly via console.log(), ReduxDevtools, etc. but the view isn't refreshing until any other action is dispatched on that particular state.

**Comment by deebloo** [comment]
@wheredoesyourmindgo what happens if you manually trigger change detection?

**Comment by wheredoesyourmindgo** [comment]
@deebloo Both ChangeDetectorRef.markForCheck() and ChangeDetectorRef.detectChanges() prior and post dispatch() call do not seem to resolve the issue.

**Comment by amcdnl** [comment]
What about inside the dispatch? I know NGRX has this similar issue.

**Comment by yarrgh** [comment]
Works if you add `cdr.detach()` to the constructor:

Example:
```typescript
export class AppComponent implements OnInit {
  @Select('app.count') count$: Observable<any>;

  modifiedCount$: Observable<any>;

  constructor(private store: Store, private cdr: ChangeDetectorRef) {
    cdr.detach();
  }

  ngOnInit(): void {
    this.modifiedCount$ = this.count$.pipe();

    this.modifiedCount$.subscribe(v => {
      if (v % 2 === 0) {
        this.store.dispatch(new Add());
        this.cdr.detectChanges();
      }
    });
  }

  onClick() {
    this.store.dispatch(new Add());
  }
}
```

**Comment by amcdnl** [comment]
@yarrgh - that only work if you detach? That is super odd!

**Comment by wheredoesyourmindgo** [comment]
Just confirming that @yarrgh 's example does work for me.

**Comment by amcdnl** [comment]
@yarrgh - What is this empty pipe for? `this.modifiedCount$ = this.count$.pipe();`

**Comment by yarrgh** [comment]
I was messing around with different rxjs operators. I later removed them but didn't clean up everything after I finally got change detection working.

**Comment by amcdnl** [comment]
@deebloo - can you take a look at this one?

**Comment by amcdnl** [comment]
I this still an issue?

**Comment by tuurbo** [comment]
Still an issue for me as of 3.0.1

**Comment by yarrgh** [comment]
I really don't see this as an issue. This case isn't really a real world example and this can easily be fixed by using pipe operators to automatically dispatch a new action if needed.

[Working Stackblitz](https://stackblitz.com/edit/ngxs-simple-yv9ecf?file=app%2Fapp.component.ts)

Example (taken from the stackblitz from OP):
```ts
@Component({
  selector: 'my-app',
  template: `
    <h1>Count is {{count$ | async}}</h1>
    <button (click)="onClick()">Click Me</button>
  `
})
export class AppComponent implements OnInit {
  count$: Observable<any>;

  constructor(private store: Store, private cdr: ChangeDetectorRef) {
    // this.cdr.detach()
  }

  ngOnInit(): void {
    this.count$ = this.store.select(x => x.app.count).pipe(
      distinctUntilChanged(),
      tap(value => {
        console.log(value);
        if (value % 2 === 0) {
          this.store.dispatch(new Add());
        }
      })
    );
  }

  onClick() {
    this.store.dispatch(new Add());
  }
}
```

**Comment by amcdnl** [comment]
Closing per @yarrgh comments.
